core/slice.go

```diff
@@ -505,61 +541,63 @@ func (sl *Slice) computePendingHeader(localPendingHeaderWithTermini types.Pendin
-	localPendingHeader, exists := sl.phCache[hash]
+	localPendingHeader, exists := sl.readPhCache(hash)
 
 	if exists {
 		combinedPendingHeader := types.CopyHeader(localPendingHeader.Header)
 		for _, i := range indices {
 			combinedPendingHeader = sl.combinePendingHeader(pendingHeader.Header, combinedPendingHeader, i, false)
 		}
-		sl.writeToPhCacheAndPickPhHead(types.PendingHeader{Header: combinedPendingHeader, Termini: localPendingHeader.Termini}, nil)
-
+		bestPh, exist := sl.readPhCache(sl.bestPhKey)
+		if !exist {
+			log.Error("BestPh Key does not exist for", "key", sl.bestPhKey)
+		}
+		oldBestPhEntropy := new(big.Int).Set(bestPh.Header.CalcPhS())
+		sl.updatePhCache(types.PendingHeader{Header: combinedPendingHeader, Termini: localPendingHeader.Termini}, false)
+		sl.pickPhHead(types.PendingHeader{Header: combinedPendingHeader, Termini: localPendingHeader.Termini}, oldBestPhEntropy)
 		return nil
 	}
 	log.Warn("no pending header found for", "terminus", hash, "pendingHeaderNumber", pendingHeader.Header.NumberArray(), "Hash", pendingHeader.Header.ParentHash(), "Termini index", terminiIndex, "indices", indices)
 	return errors.New("no pending header found in cache")
 }
 
-// writePhCache dom writes a given pendingHeaderWithTermini to the cache with the terminus used as the key.
-func (sl *Slice) writeToPhCacheAndPickPhHead(pendingHeaderWithTermini types.PendingHeader, appendTime *time.Duration) bool {
-	nodeCtx := common.NodeLocation.Context()
-	bestPh, exist := sl.phCache[sl.bestPhKey]
-	if !exist {
-		log.Error("BestPh Key does not exist for", "key", sl.bestPhKey)
-		return false
-	}
-	oldBestPhEntropy := new(big.Int).Set(bestPh.Entropy)
-
+// updatePhCache updates cache given a pendingHeaderWithTermini with the terminus used as the key.
+func (sl *Slice) updatePhCache(pendingHeaderWithTermini types.PendingHeader, inSlice bool) {
 	// Update the pendingHeader Cache
-	oldPh, exist := sl.phCache[pendingHeaderWithTermini.Termini[c_terminusIndex]]
+	oldPh, exist := sl.readPhCache(pendingHeaderWithTermini.Termini[c_terminusIndex])
 	var deepCopyPendingHeaderWithTermini types.PendingHeader
 	newPhEntropy := pendingHeaderWithTermini.Header.CalcPhS()
-	deepCopyPendingHeaderWithTermini = types.PendingHeader{Header: types.CopyHeader(pendingHeaderWithTermini.Header), Termini: pendingHeaderWithTermini.Termini, Entropy: newPhEntropy}
+	deepCopyPendingHeaderWithTermini = types.PendingHeader{Header: types.CopyHeader(pendingHeaderWithTermini.Header), Termini: pendingHeaderWithTermini.Termini}
 	deepCopyPendingHeaderWithTermini.Header.SetLocation(common.NodeLocation)
 	deepCopyPendingHeaderWithTermini.Header.SetTime(uint64(time.Now().Unix()))
 	if exist {
-		if newPhEntropy.Cmp(oldPh.Entropy) >= 0 {
-			sl.phCache[pendingHeaderWithTermini.Termini[c_terminusIndex]] = deepCopyPendingHeaderWithTermini
+		// If we are inslice we will only update the cache if the entropy is better
+		if inSlice && newPhEntropy.Cmp(oldPh.Header.CalcPhS()) >= 0 {
```

Comment:
This condition should just be `if !inSlice || newPhEntropy.Cmp(oldPh.Header.CalcPhS()) >= 0 ` and then you don't need the `else if`.